### PR TITLE
adding keycloak version of SAMLUserDetailService implement

### PR DESCRIPTION
**Pull request by khzhu**
# What? Why?
- [ ] Implements Spring Security SAMLUserDetailsService Interface to instantiate PortalUserDetails object from SAML assertion received from keycloak identity rather than from the cBioPortal database tables to centralize user control
- [ ] Creates Spring Security GrantedAuthority list based on cancer study roles included in the assertion during authentication to have a security layer sit completely on top of cBioPortal
- [ ] Uses saml.custom.userservice.class property of portal properties file to ping down SAMLUserDetailsService class used in SAML authentication to support two models: authorizing user using database tables or using keycloak at the same time

Changes proposed in this pull request:
- new sub folder keycloak under org.cbioportal.security.spring.authentication
- new class org.cbioportal.security.spring.authentication.keycloak.SAMLUserDetailsService.java

# Checks
- [ ] Runs on Heroku.
- [ ] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [ ] Follows the [Google Style Guide](https://github.com/google/styleguide).
- [ ] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)
- [ ] If this is a feature, the PR is to rc. If this is a bug fix, the PR is to
  hotfix.

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [GifGrabber](http://www.gifgrabber.com/).

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). If you are part of the
cBioPortal organization, notify the approprate team (remove inappropriate):

@cBioPortal/frontend
@cBioPortal/backend
@cBioPortal/devops

If you are not part of the cBioPortal organization look at who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them here:

### CODE REVIEW

**Comment by n1zea144** [comment]
Hi @khzhu.  I created a new branch called keycloak-support.  Please update the PR to point to that.  I haven't looked at the code in detail yet, but I noticed you make reference the Hyve in your copyright notice.

Lastly, my understanding is that this PR is to provide support for the keycloak package.  Do you plan to add more to this PR, property/configuration files, documentation, etc?

**Comment by khzhu** [comment]
I will include document on how to set up keycloak and a sample portal.properties file for your review.

**Comment by khzhu** [comment]
sorry accidentally clicked on wrong button. not mean to close the pull.

**Comment by khzhu** [comment]
[authentication_with_keycloak.docx](https://github.com/cBioPortal/cbioportal/files/906016/authentication_with_keycloak.docx)

**Comment by n1zea144** [comment]
Hi @khzhu What we really need to do is get it into the [docs](https://github.com/cBioPortal/cbioportal/tree/keycloak-support/docs) directory, so it will appear on [readthedocs](https://cbioportal.readthedocs.io/en/latest/).

Before that though, I think it would be worth going through this in more detail over a google hangout with you and @mandawilson.  I can send an email to coordinate a time.

**Comment by khzhu** [comment]
thank you, meanwhile I will create a md file in the docs directory.

**Comment by n1zea144** [comment]
If its helpful, you can look at other security related pages (like existing SAML) for a reference.

**Comment by khzhu** [comment]
commited docs/Authenticating-Users-via-keycloak-SAML.md file.

**Comment by n1zea144** [comment]
Hi @khzhu I've asked @mandawilson to look at this PR and help get it in shape so that once merged, users that desire to use keycloak  can get up to speed relatively easily.  Manda said she will have time to look at this on Monday.

One thing we will have to manage here is how much we expect a user to install keycloak versus relying on an institution system already running.  Having said that, do you know if there is an openly accessible key cloak instance somewhere for testing?

**Comment by khzhu** [comment]
Hi @n1zea144, I will discuss this with Trevor see if we can offer any helps. I got a copy of keycloak installed on my mac for developing. Keycloak is a free software and doubt there are any free providers for testing.

**Comment by n1zea144** [comment]
Hi @khzhu I suppose Manda can do that too.  I'm going to be out of town from tomorrow through next Thursday. Feel free to have a discussion with @mandawilson without me to get this moving forward.  As  I mentioned, I think we should just fill in this PR enough so that users that want to use keycloak can get up and running as easily as possible.

I now think we can assume that if a user wants to use cBioPortal support for keycloak that they have an existing keycloak instance.  I would limit the documentation to properly setting up properties on the cBioPortal side and adding studies and associating users to them on the keycloak side.

**Comment by khzhu** [comment]
Hi @n1zea144, thank you so much! I will follow up PR with Manda. I've already checked in Authenticating-Users-via-keycloak-SAML.md file to my fork/keycloak and recorded major steps I went through to get cBioPortal worked with the keycloak. I will need add more instructions on how to add studies and grant user to view those studies.

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        String userId = null;
+        List<String> userRoles = new ArrayList<String>();
+
+        // get userid and roles: iterate over attributes searching for "email" and "roles":
+        for (Attribute cAttribute : credential.getAttributes()) {
+            String attrName = cAttribute.getName();
+            log.debug("loadUserBySAML(), parsing attribute -" + attrName);
+
+
+            if (userId == null && attrName == SAML_IDP_METADATA_EMAIL_ATTR_NAME) {
```
Comment (mandawilson) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java]: String comparison should be done with equals() not ==.

File: security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                userId = credential.getAttributeAsString(cAttribute.getName());
+                log.debug(userId);
+            }
+            if (attrName == SAML_IDP_METADATA_ROLE_ATTR_NAME) {
```
Comment (mandawilson) [inline security/security-spring/src/main/java/org/cbioportal/security/spring/authentication/keycloak/SAMLUserDetailsServiceImpl.java]: String comparison should be done with equals() not ==.

File: docs/Authenticating-Users-via-keycloak-SAML.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
+# Introduction
```
Comment (mandawilson) [inline docs/Authenticating-Users-via-keycloak-SAML.md]: Please update the documentation to the version you sent me in Slack.  And add a note to that version to set the "SAML Attribute Name" to "email" to match your screenshot in the "X500 Email" email page.

**Comment by khzhu** [comment]
sorry for my late response. I was on vacation. 
Thank you for reviewing the code! khzhu/cbioportal master branch should include the updated document and fixing.

**Comment by mandawilson** [comment]
Hi @khzhu Can you update the khzhu/keycloak branch because that is the one used in this PR.  Thanks!

**Comment by khzhu** [comment]
our network is awfully slow today. It might take a while. I will let you know once I pushed the 
changes to the keycloak branch.

**Comment by mandawilson** [comment]
No problem!  Can you also do a git rebase to squash your commit into one commit?  Let me know if you have any questions.

**Comment by khzhu** [comment]
rebased and pushed to git, thank you, Manda!

**Comment by mandawilson** [comment]
Beautiful!

**Pull request merged**